File: medium/Two-Pointers/remove-duplicates-from-sorted-array-ii/remove_duplicates.py
from typing import List

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        """
        Keep at most two duplicates of each element in-place.
        Approach: Slow & Fast Two Pointers
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if len(nums) <= 2:
            return len(nums)
        
        # slow pointer points to the position to be filled
        slow = 2
        for fast in range(2, len(nums)):
            # keep element if it is not equal to the one at slow-2
            if nums[fast] != nums[slow - 2]:
                nums[slow] = nums[fast]
                slow += 1
        return slow

File: medium/Two-Pointers/remove-duplicates-from-sorted-array-ii/remove_duplicates.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        /*
         * Keep at most two duplicates in sorted array.
         * Approach: Slow & Fast Two Pointers
         * Time Complexity: O(n)
         * Space Complexity: O(1)
         */
        int n = nums.size();
        if (n <= 2) return n;
        
        int slow = 2;
        for (int fast = 2; fast < n; ++fast) {
            if (nums[fast] != nums[slow - 2]) {
                nums[slow++] = nums[fast];
            }
        }
        return slow;
    }
};
